pkg/cache/server/handler.go

```diff
@@ -0,0 +1,113 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package server
+
+import (
+	"fmt"
+	"net/http"
+	"regexp"
+	"strings"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apiserver/pkg/endpoints/handlers/responsewriters"
+	"k8s.io/apiserver/pkg/endpoints/request"
+)
+
+var (
+	shardNameRegExp = regexp.MustCompile(`^[a-z0-9-]{0,61}$`)
+
+	errorScheme = runtime.NewScheme()
+	errorCodecs = serializer.NewCodecFactory(errorScheme)
+)
+
+func init() {
+	errorScheme.AddUnversionedTypes(metav1.Unversioned,
+		&metav1.Status{},
+	)
+}
+
+// WithShardScope reads a shard name from the URL path and puts it into the context.
+// It also trims "/shards/" prefix from the URL.
+// If the path doesn't contain the shard name then a default "cache" name is assigned.
+//
+// For example:
+//
+// /shards/*/clusters/*/apis/apis.kcp.dev/v1alpha1/apiexports
+//
+// /shards/amber/clusters/*/apis/apis.kcp.dev/v1alpha1/apiexports
+//
+// /shards/sapphire/clusters/system:sapphire/apis/apis.kcp.dev/v1alpha1/apiexports
+//
+// /shards/amber/clusters/system:amber/apis/apis.kcp.dev/v1alpha1/apiexports
+func WithShardScope(handler http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		var shardName string
+		if path := req.URL.Path; strings.HasPrefix(path, "/shards/") {
+			path = strings.TrimPrefix(path, "/shards/")
+
+			i := strings.Index(path, "/")
+			if i == -1 {
+				responsewriters.ErrorNegotiated(
+					apierrors.NewBadRequest(fmt.Sprintf("unable to parse shard: no `/` found in path %s", path)),
+					errorCodecs, schema.GroupVersion{},
+					w, req)
+				return
+			}
+			shardName, path = path[:i], path[i:]
+			req.URL.Path = path
+			for i := 0; i < 2 && len(req.URL.RawPath) > 1; i++ {
```

Comment:
what's the rational of this loop? Can you describe in one or two setences?

Reply:
in comment

Reply:
copy&pase from the WithClusterScope filter but I think it is broken and not needed, opened https://github.com/kcp-dev/kcp/pull/1883